Arnold : Add width and height to `quad_light`

ericmehl opened the pull request:
This adds `width` and `height` parameters to Arnold quad lights.

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

ericmehl commented:
I think it's valid ( and needed ) to remove the `vertices` parameter in the USD quad light test in `ShaderNetworkAlgoTest` because it should be tested now by `testQuadLightVertices()`, but I want to point that out in particular for review.

ericmehl commented:
I'm not sure where the test failure is coming from, it seems specifying the vertices ourselves results in a slightly brighter pixel sampled at the center of the image.

I'm pretty confident that the default width and height are correct to what Arnold defaults to (tested in `IECoreArnoldTest.ShaderNetworkAlgoTest.testQuadLightVertices`. 

Commenting out the translation code in `IECoreArnold::ShaderNetworkAlgo::preprocessedNetwork()` does fix the problem, so I've at least confirmed it's related to the `vertices` being set.

I tried adding an `AiNodeParameterReset()` by adding `vertices` alongside `color` at https://github.com/GafferHQ/gaffer/blob/9144dd45a4021dfdd135c5d9765354b09c1948e2/src/IECoreArnold/ShaderNetworkAlgo.cpp#L204 on the off chance there is another reset problem. No luck.

I'll keep investigating, if anyone has an idea to try, the tipline is open...

johnhaddon commented on src/IECoreArnold/ShaderNetworkAlgo.cpp:
```diff
@@ -317,6 +317,40 @@ ShaderNetworkPtr preprocessedNetwork( const IECoreScene::ShaderNetwork *shaderNe
+			}
+
+			newShader->parameters()["vertices"] = new V3fVectorData( {
+				V3f( width / 2, height / 2, 0 ),
+				V3f( width / 2, -height / 2 , 0 ),
+				V3f( -width / 2, -height / 2, 0 ),
+				V3f( -width / 2, height / 2, 0 )
```
One possibility for the sampling change in the test is that our vertex ordering differs from Arnold's default :

```import arnold

arnold.AiBegin()

n = arnold.AiNodeEntryLookUp( "quad_light" )
p = arnold.AiNodeEntryLookUpParameter( n, "vertices" )
d = arnold.AiParamGetDefault( p )
a = d.contents.ARRAY
for i in range( 0, arnold.AiArrayGetNumElements( a ) ) :
	print( i, ":", arnold.AiArrayGetVec( a, i ).x, arnold.AiArrayGetVec( a, i ).y, arnold.AiArrayGetVec( a, i ).z )

arnold.AiEnd()
```

Outputs :

```
0 : -1.0 -1.0 0.0
1 : -1.0 1.0 0.0
2 : 1.0 1.0 0.0
3 : 1.0 -1.0 0.0
```

ericmehl replied:
That definitely fixes the test, and I think we need to adhere to that ordering to keep from reorienting everyone's quad light textures when they update to the version this ends up in.

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/5169/commits/7b28509c04775978ef866e9255b7ce3d92fbb516

johnhaddon commented on src/IECoreArnold/ShaderNetworkAlgo.cpp:
```diff
@@ -1003,14 +1037,6 @@ void IECoreArnold::ShaderNetworkAlgo::convertUSDShaders( ShaderNetwork *shaderNe
 		{
 			newShader = new Shader( "quad_light", "ai:light" );
 			transferUSDLightParameters( shaderNetwork, handle, shader.get(), newShader.get() );
-			const float width = parameterValue( shader.get(), g_widthParameter, 1.0f );
```
We need to transfer the width and height to `newShader` using `transferUSDParameter()`, bearing in mind that the USD defaults are different to the Arnold ones (would be good to add test coverage for this case). If we do need to change the vertex order though, we need to be wary of how this impacts the orientation of any textures applied to the light - it's possible that we actually want to keep on defining vertices here, if the USD ones require a different order to the default Arnold ones.

ericmehl replied:
> it's possible that we actually want to keep on defining vertices here

Yes, I think we do need to keep defining the vertices for USD lights. To get the proper orientation for what I can gather from the spec is the right layout we need to start from a different corner point when converting to Arnold, and add a `visualiserOrientation` for GL.

Those are both done in https://github.com/GafferHQ/gaffer/pull/5169/commits/7b28509c04775978ef866e9255b7ce3d92fbb516

johnhaddon replied:
Did you manage to get a practical reference for a textured USD quad light? If not, maybe we can get one out of Houdini, @murraystevenson?

johnhaddon commented on startup/GafferScene/arnoldLights.py:
```diff
@@ -64,6 +64,8 @@
 Gaffer.Metadata.registerValue( "ai:light:quad_light", "intensityParameter", "intensity" )
 Gaffer.Metadata.registerValue( "ai:light:quad_light", "exposureParameter", "exposure" )
 Gaffer.Metadata.registerValue( "ai:light:quad_light", "colorParameter", "color" )
+Gaffer.Metadata.registerValue( "ai:light:quad_light", "widthParameter", "height" )  # Width and height are reversed because of `visualiserOrientation`
```
This seems fishy - perhaps a sign that we should change vertex ordering, or that `visualiserOrientation` was wrong?

ericmehl replied:
I suppose there's another possible approach, which is to leave the `visualiserOrientation` as it is with the 90 degree rotation, and also have `widthParameter = "width"` and `heightParameter = "height"`.

![image](https://user-images.githubusercontent.com/7873467/222257495-a03d97d0-194e-4441-8d49-0d70920eb2c5.png)

That scheme with a width of 4 and a height of 2 gives the light above.

To me, that has the x and y axes reversed and is what I've been trying to correct for. But from the point of view of the image, the axes are not reversed. Maybe people who use images on the quad light are accustomed to rotating it?

johnhaddon replied:
The job of the visualiser is to match what Arnold will do, even if what Arnold does is a bit odd. The old `visualiserOrientation` accounted for the fact that Arnold would rotate the texture 90 degrees by default, and the new visualisation no longer matches Arnold's behaviour. Unfortunately, the new visualisation _does_ seem to match for width and height, so I think we do need something like the `uvOrientation` you suggested, to separate the texture orientation from the geometry orientation.

johnhaddon reviewed (changes requested):
Thanks Eric - comments inline as usual.

ericmehl commented on python/GafferArnoldUI/ArnoldShaderUI.py:
```diff
@@ -392,6 +392,16 @@ def addActivator( activator ) :
+	__metadata["quad_light.parameters.width"]["layout:index"] = 0
+	__metadata["quad_light.parameters.height"]["layout:index"] = 1
+	GafferSceneUI.LightEditor.registerParameter( "ai:light", "width", "Shape" )
+	GafferSceneUI.LightEditor.registerParameter( "ai:light", "height", "Shape" )
```
That was an oversight on my part, fixed in https://github.com/GafferHQ/gaffer/pull/5169/commits/b572c92b4e793f70cfd0a89d66172f46ee71e6ea

ericmehl commented on src/GafferArnold/ArnoldShader.cpp:
```diff
@@ -171,8 +173,44 @@ void ArnoldShader::loadShader( const std::string &shaderName, bool keepExistingV
+		else
+		{
+			parametersPlug->addChild(
+				new FloatPlug( "width", Gaffer::Plug::Direction::In, 2.f, 0.f, limits<float>::max() )
```
Fixed in https://github.com/GafferHQ/gaffer/pull/5169/commits/b24b5b14ae84b259738ddd6d276fcd875a6a313e

ericmehl commented on src/GafferArnold/ArnoldShader.cpp:
```diff
@@ -171,8 +173,44 @@ void ArnoldShader::loadShader( const std::string &shaderName, bool keepExistingV
+		// they are not needed. Hold on to them if they exist and add them back in
+		// after `setupPlugs()`.
+		existingWidthPlug = parametersPlug->getChild<FloatPlug>( "width" );
+		existingHeightPlug = parametersPlug->getChild<FloatPlug>( "height" );
```
That is nicely simple, changed in https://github.com/GafferHQ/gaffer/pull/5169/commits/ba0955ef50c98cb4468d22c7cca64cfd2307ea26

ericmehl commented on src/IECoreArnold/ShaderNetworkAlgo.cpp:
```diff
@@ -317,6 +317,40 @@ ShaderNetworkPtr preprocessedNetwork( const IECoreScene::ShaderNetwork *shaderNe
+				V3f( -width / 2, -height / 2, 0 ),
+				V3f( -width / 2, height / 2, 0 )
+			} );
+
```
Fixed in https://github.com/GafferHQ/gaffer/pull/5169/commits/902809f8797a6f26b561271d2882a5a85962a988

ericmehl commented on python/IECoreArnoldTest/ShaderNetworkAlgoTest.py:
```diff
@@ -822,14 +897,7 @@ def expectedLightParameters( parameters ) :
-							imath.V3f( -10, -30, 0 ),
-							imath.V3f( -10, 30, 0 ),
-						] )
-					} )
```
Added back in https://github.com/GafferHQ/gaffer/pull/5169/commits/7b28509c04775978ef866e9255b7ce3d92fbb516

ericmehl commented:
I have the comments addressed, sans the GL orientation issue for Arnold `quad_light`.

johnhaddon commented on src/GafferArnold/ArnoldShader.cpp:
```diff
@@ -43,6 +43,7 @@
 
 #include "Gaffer/CompoundNumericPlug.h"
 #include "Gaffer/NumericPlug.h"
+#include "Gaffer/PlugAlgo.h"
```
We don't need this include any more.

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/5169/commits/b7b54bcc7fa8bc97c318ecbb5b8b461e8e7ac31b

johnhaddon commented on src/IECoreArnold/ShaderNetworkAlgo.cpp:
```diff
@@ -311,12 +311,86 @@ void resetNode( AtNode *node )
+				V3f( width / 2, -height / 2, 0 )
+			} );
+
+			auto it = newShader->parameters().find( "width" );
+			if( it != newShader->parameters().end() )
+			{
+				newShader->parameters().erase( it );
+			}
```
This can be a one-liner : `newShader->parameters().erase( "width" );`

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/5169/commits/d9eb6410c95420a25f19d2da84fcb3fd2e4f26be

ericmehl commented:
I took care of the two small fixes, and added a second commit where I add the `uvOrientation` metadata.

I tested the texture orientation in Houdini, and it turns out their implementation is flopped vs ours - our macaw points to "stage-right", and theirs points stage-left. The source image viewed in an image viewer has the macaw facing the right.

I think the Houdini way is not correct though, it seems like a strange interpretation of the USD description of
```
The rectangle is centered in the XY plane and emits light along the -Z axis. The rectangle is 1 unit in length in the X and Y axis. In the default position, a texture file's min coordinates should be at (+X, +Y) and max coordinates at (-X, -Y).
```

to consider the texture origin at the top-right.

Maybe @murraystevenson could double-check that my Houdini setup is correct? The image I get looking at a reflective plane is :
![image](https://user-images.githubusercontent.com/7873467/222574602-8e117d15-16b2-4e82-831a-0afda3bf029e.png)

murraystevenson commented:
I ran some tests from Maya and Houdini and the macaw facing "stage-right" when looking at the light does seem preferable. That's the result you get out of MtoA's Arnold Render View, and when using Arnold in Maya's viewport. 

Houdini seems _variable_. Arnold in the viewport and Arnold renders via `husk` draw upside-down facing stage-left, but the same scene rendered via [sick](https://help.autodesk.com/view/ARNOL/ENU/?guid=arnold_for_houdini_solaris_ah_usd_render_rop_html) draws right-way-up and facing stage-right. Karma renders appear to be right-way-up and facing stage-left.

Moving to Arnold in usdview, a USD file output from Houdini draws upside-down facing stage-left, but the equivalent light output from arnold-usd in Maya draws right-way-up facing stage-right. The difference there is the Maya arnold-usd export authors `vector3f[] primvars:arnold:vertices = [(1, -1, 0), (-1, -1, 0), (-1, 1, 0), (1, 1, 0)]` on the light.

:upside_down_face: 

[testRectLightMayaAndHoudini.zip](https://github.com/GafferHQ/gaffer/files/10877232/testRectLightMayaAndHoudini.zip)

johnhaddon reviewed (commented):
I've noted a couple of tiny additional things that it'd be worth sorting before we merge this...otherwise I think this is good to squash down...

ericmehl commented:
Thanks for doing all that research Murray! Who knew putting an image on a rectangle would lead to so many different approaches...

> the Maya arnold-usd export authors `vector3f[] primvars:arnold:vertices = [(1, -1, 0), (-1, -1, 0), (-1, 1, 0), (1, 1, 0)]`

For USD quad lights to Arnold, that's what we author as well, so at least there's one point of agreement!

This PR uses `[( -, -, 0 ), ( -, +, 0 ), ( +, +, 0 ), ( +, -, 0 )]` ordering for native Arnold quad lights to align with their default which I think is best to keep for backwards compatibility.

Given the state of the other renderers, the fact that we agree with at least a few of them and that we're using at least a viable interpretation of the USD spec, I'm inclined to keep the current implementation. But I'm open to opposing ideas as well.

johnhaddon commented:
This all seems to be behaving as it should now, which is great. Thanks Eric! But it comes at the expense of us layering corrections on top of corrections (through necessity, because we need to preserve compatibility). So I want to at least take the opportunity to get things straight in my head by documenting what I think is going on - correct me where I'm wrong please.

1. The docs for UsdLuxRectLight define the ideal mapping of a texture to a quad light. The mapping means that the image is orientated naturally from the point of view of the illuminated surface. Put a rect light at the origin. Put a vertical plane a few units back in Z. Look at the light from the plane. You see the texture oriented naturally just as you would if it were in an image viewer. Makes perfect sense.
2. But Arnold's default `quad_light.vertices` value means that the texture will actually be rotated through 90 degrees. If we had noticed this from the outset, we could have corrected it, but for backwards compatibility we need to stick with it for now.
3. `IECoreGL::QuadPrimitive` defines a perfectly sensible set of UVs, which orient the image correctly for a camera looking at it down -ve Z. This is not what we need in the visualiser, because this camera is looking in the opposite direction to the "fly on the wall" from 1. We would need to flop the U values to correct for that.
4. But instead, `StandardLightVisualiser::quadSurface()` [flips the geometry in Y](https://github.com/GafferHQ/gaffer/blob/main/src/GafferSceneUI/StandardLightVisualiser.cpp#L980-L982), for reasons unknown.
5. To compensate for this mess, we have been using `visualiserOrientation` to allow an arbitrary 4x4 corrective transform to be applied to the geometry. But if you use that to correct Arnold's texture orientation, you have changed the axes in which `width` and `height` are applied. So we have instead introduced a new `uvOrientation` corrective which just changes the texture mapping without affecting the geometry.
6. In a sane world, for quad lights at least, we wouldn't need any correctives at all. A simple flop in U would have done it.

ericmehl commented:
Yes, that is all correct, thanks for laying it out nice and clear. No doubt someone in the future will see all this and wonder why we made it so complicated.

I fixed up the latest couple of comments and squashed down, so hopefully this is ready to go.

ericmehl commented on src/IECoreArnold/ShaderNetworkAlgo.cpp:
```diff
@@ -1014,10 +1042,10 @@ void IECoreArnold::ShaderNetworkAlgo::convertUSDShaders( ShaderNetwork *shaderNe
 			const float width = parameterValue( shader.get(), g_widthParameter, 1.0f );
 			const float height = parameterValue( shader.get(), g_heightParameter, 1.0f );
 			newShader->parameters()[g_verticesParameter] = new V3fVectorData( {
-				V3f( width / 2, height / 2, 0 ),
 				V3f( width / 2, -height / 2, 0 ),
 				V3f( -width / 2, -height / 2, 0 ),
-				V3f( -width / 2, height / 2, 0 )
+				V3f( -width / 2, height / 2, 0 ),
+				V3f( width / 2, height / 2, 0 )
```
Fixed and squashed into https://github.com/GafferHQ/gaffer/pull/5169/commits/cfd1fbc9ebc93048b94bfa2d701652d3e1c94b65

ericmehl commented on src/GafferSceneUI/StandardLightVisualiser.cpp:
```diff
@@ -525,6 +551,12 @@ Visualisations StandardLightVisualiser::visualise( const IECore::InternedString
+		ConstM33fDataPtr uvOrientation = Metadata::value<M33fData>( metadataTarget, "uvOrientation" );
+		if( !uvOrientation )
+		{
+			uvOrientation = new M33fData( M33f() );
```
Fixed and squashed into https://github.com/GafferHQ/gaffer/pull/5169/commits/9d32e96a15f779ba3d68f9f1ce9bdef9db9d69ae
